### PR TITLE
DrivAerML: 4L/384d + T_max=40 — finer cosine resolution

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

T_max=30 showed oscillation at 384d (epoch 141=10.2%, epoch 144=5.7%), suggesting the schedule may be slightly aggressive. T_max=40 sits between the current T_max=30 and T_max=50 (being tested in parallel). This explores a finer T_max resolution point — if T_max=30 overshoots and T_max=50 undershoots, T_max=40 may be the sweet spot for stable convergence at 384d width.

## Instructions

Run the following command exactly (one change from base: `--cosine-t-max 40`):

```bash
cd target/icml2026 && python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 40 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 384 \
  --model-heads 6 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --agent ymir \
  --wandb-name "ymir/drivaerml-4L384d-tmax40" \
  --wandb-group "drivaerml-4L384d-variants"
```

**Single change from base:** `--cosine-t-max 40` (all other hyperparameters identical to PR #2602)

## Baseline

Current DrivAerML best (PR #2602):
- **val_primary/surface_rel_l2_pct:** 5.73% (epoch 144)
- **Baseline W&B run:** 7ogfs7ph (151 epochs, 180-min budget)
- **Architecture:** 4L/384d/6H + Fourier + no-EMA + AdamW lr=5e-4 + T_max=30
- **Reproduce baseline:** `cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 384 --model-heads 6 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`
- **External target:** <3.71% (AB-UPT)

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2687 because ADVISOR: Closing — 4L/384d+T_max=40 superseded by 4L/320d baseline. Future DrivAerML experiments should use 4L/320d..

**Comment by tcapelle** [comment]
STUDENT ymir:

## Results

**Experiment:** DrivAerML 4L/384d + T_max=40 (Fourier, no-EMA)

| Metric | Baseline (PR #2602, 4L/384d T_max=30, 151ep) | This Run |
|--------|-----------------------------------------------|----------|
| val_primary/surface_rel_l2_pct | 5.73% | **4.84%** |
| test_primary/surface_rel_l2_pct | — | 7.39% |
| Epochs | 151 | 266 |

**Best val: 4.84%** at epoch 259 — beats baseline by 0.89%

### Command
```bash
CUDA_VISIBLE_DEVICES=0 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 40 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 384 --model-heads 6 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --agent ymir --wandb-name "ymir/drivaerml-4L384d-tmax40" --wandb-group "drivaerml-4L384d-variants"
```

### W&B
- Run ID: `cngg4zak`
- URL: https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/cngg4zak

### Peak Memory
- 8 GPUs used (DDP), batch_size=1, 50k surface points

### Cycle-bottom progression (T_max=40)
Epoch 20: 21.21% → 40: 13.91% → 60: 9.90% → 80: 8.46% → 100: 7.66% → 120: 6.70% → 140: 6.07% → 160: 6.64% → 180: 5.83% → 200: 5.62% → 220: 5.25% → 240: 5.39% → 260: 4.96%

### What happened
The 4L/384d model with T_max=40 performed very well. It reached sub-5% validation (4.84%) by epoch 259, beating the previous 5.73% baseline by 0.89 percentage points. The cosine annealing warm restarts showed consistent improvement across cycles, with each cycle bottom generally improving over the last. The model showed some increasing noise/variance in later cycles (epochs 200+) which is typical as the model approaches convergence.

The test metric (7.39%) is notably higher than the best val (4.84%), indicating some val-test generalization gap. This is worth investigating in future experiments.

**Note:** PR was closed by advisor — 4L/384d+T_max=40 superseded by 4L/320d baseline. Results posted for the record.

### Suggested follow-ups
- The 4L/320d architecture (advisor's new baseline) may reach similar results with less compute due to fewer parameters
- The val-test generalization gap (4.84% vs 7.39%) suggests possible overfitting — regularization (dropout, weight decay tuning) or data augmentation could help
- T_max=30 or T_max=20 with the 4L/320d architecture may converge faster to similar results
- The model was still improving at timeout — longer runs could push further below 4.84%
